server/lib/tuist/ingestion/buffer.ex

```diff
@@ -0,0 +1,163 @@
+    flush_interval_ms = opts[:flush_interval_ms] || default_flush_interval_ms()
+
+    Process.flag(:trap_exit, true)
+    timer = Process.send_after(self(), :tick, flush_interval_ms)
+
+    {:ok,
+     %{
+       buffer: buffer,
+       timer: timer,
+       name: Keyword.fetch!(opts, :name),
+       insert_sql: Keyword.fetch!(opts, :insert_sql),
+       insert_opts: Keyword.fetch!(opts, :insert_opts),
+       header: Keyword.fetch!(opts, :header),
+       buffer_size: IO.iodata_length(buffer),
+       max_buffer_size: max_buffer_size,
+       flush_interval_ms: flush_interval_ms
+     }}
+  end
+
+  @impl true
+  def handle_cast({:insert, row_binary}, state) do
+    state = %{
+      state
+      | buffer: [state.buffer | row_binary],
+        buffer_size: state.buffer_size + IO.iodata_length(row_binary)
+    }
+
+    if state.buffer_size >= state.max_buffer_size do
+      Logger.notice("#{state.name} buffer full, flushing to ClickHouse")
+      Process.cancel_timer(state.timer)
+      do_flush(state)
+      new_timer = Process.send_after(self(), :tick, state.flush_interval_ms)
+      {:noreply, %{state | buffer: [], timer: new_timer, buffer_size: 0}}
+    else
+      {:noreply, state}
+    end
+  end
+
+  @impl true
+  def handle_info(:tick, state) do
+    do_flush(state)
+    timer = Process.send_after(self(), :tick, state.flush_interval_ms)
+    {:noreply, %{state | buffer: [], buffer_size: 0, timer: timer}}
+  end
+
+  @impl true
+  def handle_call(:flush, _from, state) do
+    %{timer: timer, flush_interval_ms: flush_interval_ms} = state
+    Process.cancel_timer(timer)
+    do_flush(state)
+    new_timer = Process.send_after(self(), :tick, flush_interval_ms)
+    {:reply, :ok, %{state | buffer: [], buffer_size: 0, timer: new_timer}}
+  end
+
+  @impl true
+  def terminate(_reason, %{name: name} = state) do
+    Logger.notice("Flushing #{name} buffer before shutdown...")
+    do_flush(state)
+  end
+
+  defp do_flush(state) do
+    %{
+      buffer: buffer,
+      buffer_size: buffer_size,
+      insert_opts: insert_opts,
+      insert_sql: insert_sql,
+      header: header,
+      name: name
+    } = state
+
+    case buffer do
+      [] ->
+        nil
+
+      _not_empty ->
+        Logger.notice("Flushing #{buffer_size} byte(s) RowBinary from #{name}")
+        IngestRepo.query!(insert_sql, [header | buffer], insert_opts)
+    end
+  end
+
+  defp default_flush_interval_ms do
+    Keyword.fetch!(Application.get_env(:tuist, IngestRepo), :flush_interval_ms)
+  end
+
+  defp default_max_buffer_size do
+    Keyword.fetch!(Application.get_env(:tuist, IngestRepo), :max_buffer_size)
+  end
+
+  @doc """
+  Prepares a few buffer parameters for a given schema at compile time.
+
+  The `RowBinaryWithNamesAndTypes` format expects a header row with the column names that we can generate before the runtime, as well as the actual SQL query to insert the data.
+  """
+  def compile_time_prepare(schema) do
```

Comment:
I'd extract this piece into a module that you can include with `use` into an Ecto schema.